feat(editor): updating editor name when modifying file studies

mehdiwahada opened the pull request:
I have implemented a method that is updating the editor's name when the user is modifying studies. The method is used inside the add_commands from RawStudyInterface.

It is a continuity of a task where the editor field was changed when a study is imported or created.

You can test it by creating a study and modify it (you can add / update comments, add areas, update their metadatas and delete them -> example in test_integration.py)

mehdiwahada committed:
feat(editor): finished update_editor_name method and integration testing

mehdiwahada committed:
feat(editor): corrected a client.delete error

mehdiwahada committed:
feat(editor): finished integration test
next : database modification

mehdiwahada committed:
feat(editor): replaced get_user_id by get_user_name and placed a study_service in study_interfaces

mehdiwahada committed:
feat(editor): finished

mehdiwahada committed:
feat(editor): finished database editor_name changing methods and wip on unit testing

mehdiwahada committed:
feat(editor): finished db variant study testing, wip on db raw study testing

mehdiwahada committed:
feat(editor): finished database editor field unit testing

mehdiwahada committed:
feat(editor): finished database editor field unit testing

mehdiwahada committed:
feat(editor): finished database editor field unit testing

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -117,6 +117,26 @@ def __init__(
         CommandMatrixUsageProvider(variant_study_repo=repository, command_factory=command_factory)
 
+    def _update_editor(self, study: VariantStudy) -> None:
+        user_id = get_user_id()
```
you can use `require_user().id` instead, and remove the if condition below together with the cast.
We must have a logged in user when we get here.

sylvlecl replied:
Actually even better: we should use `get_user_impersonator()` instead.

The current user id may be the ID of a token: using the "impersonator" ensures that we get the ID of the real user, not the ID of a token which belongs to a real user.

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -117,6 +117,26 @@ def __init__(
+
+        if user_id:
+            user_name = self._get_user_name_from_id(int(user_id))
+            if user_name:
```
we don't need to check this

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -117,6 +117,26 @@ def __init__(
+            user_name = self._get_user_name_from_id(int(user_id))
+
+            if user_name:
+                study_db.additional_data.editor = user_name
```
You already did that in the other method, do we really need both ? 
Can't we just add the save in the other method ?

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -236,6 +256,8 @@ def append_commands(self, study_id: str, commands: List[CommandDTO]) -> List[str
         ]
         study.commands.extend(new_commands)
+        self._update_editor(study)
+        self._update_db_study_editor(study.id)
```
Seems to me we can remove those duplicate method calls and only keep the `_update_editor`

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
About the handling of editor for variants:
now we correctly update the field in database, good thing.

But, we don't update the `study.(antares` file when we generate the corresponding snapshots.
This should be done in the `update_antares_info` method and for consistency sake we should also read the editor at the start of the generation in `snapshot_generator` module, method `_read_additional_data`.


Also, it seems that in variant creation, we don't correctly set the editor name from the parent study (see line 670).

sylvlecl commented on antarest/study/service.py:
```diff
@@ -419,6 +424,18 @@ def add_commands(self, commands: Sequence[ICommand], listener: Optional[ICommand
             data = FileStudyTreeConfigDTO.from_build_config(file_study.config).model_dump()
             update_cache(self._raw_study_service.cache, study.id, data)
         self._variant_study_service.on_parent_change(study.id)
+        jwt_user = get_current_user()
```
As for variants, we should use `get_user_impersonator` instead here.

Also, you should extract all that part in a dedicated `update_editor` method so that this method remains understandable

sylvlecl commented on tests/integration/test_integration.py:
```diff
@@ -1890,3 +1890,122 @@ def test_links_deletion_with_binding_constraints(
     assert res.status_code == 200, res.json()
+
+
+def test_update_with_editor(client: TestClient, admin_access_token: str):
```
We should do the same test with a variant study, you can probably do a parameterized test for this: maybe you can use the `PreparerProxy` as in other tests

sylvlecl commented on tests/variantstudy/model/test_variant_model.py:
```diff
@@ -312,3 +313,75 @@ def check_orm_operations(orm_execute_state):
+        assert study is not None
+        assert study.id == saved_id
+        assert study.parent_id == root_study_id
+        assert study.additional_data.author == "john.doe"
```
we should have an assertion on the editor too

sylvlecl commented on tests/integration/test_integration.py:
```diff
@@ -1890,3 +1890,122 @@ def test_links_deletion_with_binding_constraints(
+    group_id = res.json()["id"]
+
+    client.post("/v1/users", json={"name": "creator_2", "password": "password123"})
+    client.post("/v1/users", json={"name": "editor_2", "password": "password456"})
```
we should have a test with a "bot" user too (a token associated to a user), to check that the editor is set to the impersonator (the associated user, not the token)

sylvlecl reviewed (changes requested):
A few changes requested to:
- handle editor in places where it's missing (variant files, variant creation)
- improve code quality (remove some duplication ...)
- ensure that the editor is correctly set to the real user, even when a "bot" token is used for authentication

mehdiwahada committed:
feat(study_editor): taking reviews into account

mehdiwahada committed:
feat(study_editor): finished #1

mehdiwahada committed:
feat(study_editor): non working tests :
- test_edit_study_with_command (error at service.py:line 435 -> Mock object does not support item assignment)
- test_sta_mini_copy (sqlite3.ProgrammingError : object created in a thread can only be used in that same thread)

mehdiwahada committed:
feat(study_editor): readapting unit and integration tests

mehdiwahada committed:
feat(study_editor): readapting unit and integration tests

sylvlecl committed:
fix(tests): in-memory sqlite needs staticpool to work in multithread env

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl commented on tests/study/test_repository.py:
```diff
@@ -1250,3 +1250,31 @@ def test_count_studies__general_case(
     if expected_ids is not None:
         assert count == len(expected_ids)
+
+@with_db_context
```
It does not test anything, I will remove it

sylvlecl commented on tests/variantstudy/model/test_variant_model.py:
```diff
@@ -312,3 +313,76 @@ def check_orm_operations(orm_execute_state):
+        assert study.id == saved_id
+        assert study.parent_id == root_study_id
+        assert study.additional_data.author == "john.doe"
+        assert study.additional_data.editor == "Unknown"
```
Here the editor should be the one who created the variant study

sylvlecl commented on tests/variantstudy/model/test_variant_model.py:
```diff
@@ -312,3 +313,76 @@ def check_orm_operations(orm_execute_state):
+
+        # creating area by the author, making him the editor of the study
+        command_1 = CommandDTO(action="create_area", args={"area_name": "area_be"}, study_version=study_version)
+        command_2 = CommandDTO(action="create_area", args={"area_name": "area_fr"}, study_version=study_version)
```
to keep the test simple we can have only one command

sylvlecl commented on tests/study/test_repository.py:
```diff
@@ -1250,3 +1250,31 @@ def test_count_studies__general_case(
+    db_session: Session,
+) -> None:
+    test_workspace = "test-repository"
+    #icache: Mock = Mock(spec=ICache)
```
comments to be removed

sylvlecl commented on antarest/study/service.py:
```diff
@@ -419,6 +424,21 @@ def add_commands(self, commands: Sequence[ICommand], listener: Optional[ICommand
         self._variant_study_service.on_parent_change(study.id)
+        self.update_editor(file_study)
+
+    def update_editor(self, file_study: FileStudy) -> None:
```
we should make it private

sylvlecl commented on antarest/study/service.py:
```diff
@@ -419,6 +424,21 @@ def add_commands(self, commands: Sequence[ICommand], listener: Optional[ICommand
+
+    def update_editor(self, file_study: FileStudy) -> None:
+        jwt_user = get_user_impersonator()
+        if jwt_user:
```
we should remove the check, and rename the variable, it's not anymore a JWT user

sylvlecl commented on antarest/study/service.py:
```diff
@@ -419,6 +424,21 @@ def add_commands(self, commands: Sequence[ICommand], listener: Optional[ICommand
+                file_study.tree.save(study_antares, ["study", "antares"])
+                study_db = self._repository.get(self._study.id)
+
+                if study_db and study_db.additional_data:
```
- we should raise if study_db does not exist
- if additional_data does not exist, we should initialize it

sylvlecl commented on antarest/study/storage/utils.py:
```diff
@@ -83,20 +83,24 @@ def update_antares_info(metadata: Study, study_tree: FileStudyTree, update_autho
     if metadata.additional_data:
         author = metadata.additional_data.author
+        editor = metadata.additional_data.editor
+        # if metadata.additional_data.editor is None:
```
comment to be removed

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -344,6 +350,8 @@ def update_command(self, study_id: str, command_id: str, command: CommandDTO) ->
             study.commands[index].command = validated_commands[0].action
             study.commands[index].args = to_json_string(validated_commands[0].args)
+            self._update_editor(study)
+            # self._update_db_study_editor(study.id)
```
comments to be removed

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -117,18 +116,15 @@ def __init__(
+        user_id = get_user_impersonator()
+        study_db = self.repository.get(study.id)
+
+        if user_id and study_db:
```
user_id is never None, we should remove the check

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -117,18 +116,15 @@ def __init__(
-        return str(user_obj.name)
+    def _update_editor(self, study: VariantStudy) -> None:
+        user_id = get_user_impersonator()
+        study_db = self.repository.get(study.id)
```
actually we already have the study, we don't need to get it again

sylvlecl reviewed (commented):
Some comments that I will handle tomorrow

sylvlecl committed:
fix editor on variant creation

Also various code cleanup

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
add copy tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl commented:
I just rebased on release branch.

sylvlecl reviewed (dismissed):
Thanks!

sylvlecl committed:
remove duplicate condition

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>
